tracing-subscriber/src/fmt/format/mod.rs

```diff
@@ -520,41 +563,59 @@ impl<'a, N> fmt::Display for FmtCtx<'a, N> {
     }
 }
 
-struct FullCtx<'a, N> {
-    ctx: &'a span::Context<'a, N>,
+pub struct FullCtx<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    ctx: &'a FmtContext<'a, S, N>,
     #[cfg(feature = "ansi")]
     ansi: bool,
 }
 
-impl<'a, N: 'a> FullCtx<'a, N> {
+impl<'a, S, N: 'a> FullCtx<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
     #[cfg(feature = "ansi")]
-    pub(crate) fn new(ctx: &'a span::Context<'a, N>, ansi: bool) -> Self {
+    pub(crate) fn new(ctx: &'a FmtContext<'a, S, N>, ansi: bool) -> Self {
         Self { ctx, ansi }
     }
 
     #[cfg(not(feature = "ansi"))]
-    pub(crate) fn new(ctx: &'a span::Context<'a, N>) -> Self {
+    pub(crate) fn new(ctx: &'a FmtContext<'a, S, N>) -> Self {
         Self { ctx }
     }
 }
 
 #[cfg(feature = "ansi")]
-impl<'a, N> fmt::Display for FullCtx<'a, N> {
+impl<'a, S, N> fmt::Display for FullCtx<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         let mut seen = false;
         let style = if self.ansi {
             Style::new().bold()
         } else {
             Style::new()
         };
-        self.ctx.visit_spans(|_, span| {
-            write!(f, "{}", style.paint(span.name()))?;
-
+        self.ctx.visit_spans(|id| {
+            let span = self.ctx.ctx.span(id);
+            if let Some(metadata) = self.ctx.ctx.metadata(id) {
```

Comment:
this could be `span.metadata()` rather than having to do a second roundtrip through the lookup process?

Reply:
Yep, removed that indirection.